File: Wikipedia/Code/PageHistoryFetcher.m
//  Created by Monte Hurd on 10/9/14.
//  Copyright (c) 2014 Wikimedia Foundation. Provided under MIT-style license; please copy and modify!

#import "PageHistoryFetcher.h"
#import "AFHTTPSessionManager.h"
#import "MWNetworkActivityIndicatorManager.h"
#import "SessionSingleton.h"
#import "NSString+WMFExtras.h"
#import "NSObject+WMFExtras.h"
#import "NSDate+Utilities.h"
#import "MediaWikiKit.h"
#import "WMFRevision.h"
#import <Mantle/Mantle.h>

@implementation PageHistoryFetcher

- (instancetype)initAndFetchHistoryForTitle:(MWKTitle*)title
                                withManager:(AFHTTPSessionManager*)manager
                         thenNotifyDelegate:(id <FetchFinishedDelegate>)delegate {
    self = [super init];
    if (self) {
        self.fetchFinishedDelegate = delegate;
        [self fetchPageHistoryForTitle:title withManager:manager];
    }
    return self;
}

- (void)fetchPageHistoryForTitle:(MWKTitle*)title
                     withManager:(AFHTTPSessionManager*)manager {
    NSURL* url = [[SessionSingleton sharedInstance] urlForLanguage:title.site.language];

    NSDictionary* params = [self getParamsForTitle:title];

    [[MWNetworkActivityIndicatorManager sharedManager] push];

    [manager GET:url.absoluteString parameters:params progress:NULL success:^(NSURLSessionDataTask* operation, id responseObject) {
        //NSLog(@"JSON: %@", responseObject);
        [[MWNetworkActivityIndicatorManager sharedManager] pop];

        // Fake out an error if non-dictionary response received.
        if (![responseObject isDict]) {
            responseObject = @{@"error": @{@"info": @"History not found."}};
        }

        //NSLog(@"PAGE HISTORY DATA RETRIEVED = %@", responseObject);

        // Handle case where response is received, but API reports error.
        // Uncomment @"rvdiffto": @(-1) in the parameters to force API error for testing.
        NSError* error = nil;
        if (responseObject[@"error"]) {
            NSMutableDictionary* errorDict = [responseObject[@"error"] mutableCopy];
            errorDict[NSLocalizedDescriptionKey] = errorDict[@"info"];
            error = [NSError errorWithDomain:@"Page History Fetcher" code:001 userInfo:errorDict];
        }

        NSArray* output = @[];
        if (!error) {
            output = [self getSanitizedResponse:responseObject];
        }

        [self finishWithError:error
                  fetchedData:output];
    } failure:^(NSURLSessionDataTask* operation, NSError* error) {
        //NSLog(@"PAGE HISTORY FAIL = %@", error);

        [[MWNetworkActivityIndicatorManager sharedManager] pop];

        [self finishWithError:error
                  fetchedData:nil];
    }];
}

- (NSDictionary*)getParamsForTitle:(MWKTitle*)title {
    NSMutableDictionary* params = @{
        @"action": @"query",
        @"prop": @"revisions",
        @"rvprop": @"ids|timestamp|user|size|parsedcomment",
        @"rvlimit": @51,
        @"rvdir": @"older",
        @"titles": title.text,
        @"continue": @"",
        @"format": @"json"
        //,@"rvdiffto": @(-1) // Add this to fake out "error" api response.
    }.mutableCopy;
    return params;
}

- (NSArray*)getSanitizedResponse:(NSDictionary*)rawResponse {
    NSMutableDictionary* revisionsByDay = @{}.mutableCopy;

    if (rawResponse.count > 0) {
        NSDictionary* pages = rawResponse[@"query"][@"pages"];
        if (pages) {
            for (NSDictionary* page in pages) {
                NSArray<WMFRevision*>* revs = [[MTLJSONAdapter arrayTransformerWithModelClass:[WMFRevision class]] transformedValue:pages[page][@"revisions"]];
                
                WMFRevision* earliestRevision = revs.lastObject;
                if (earliestRevision.parentID == 0) {
                    earliestRevision.revisionSize = earliestRevision.articleSizeAtRevision;
                    [self updateRevisionsByDay:revisionsByDay withRevision:earliestRevision];
                }
                
                for (NSInteger i = revs.count - 2; i >= 0; i--) {
                    WMFRevision* previous = revs[i + 1];
                    WMFRevision* current = revs[i];
                    current.revisionSize = current.articleSizeAtRevision - previous.articleSizeAtRevision;
                    [self updateRevisionsByDay:revisionsByDay withRevision:current];
                }
            }
        }
    }
    
    NSArray * sortedKeys = [[revisionsByDay allKeys] sortedArrayUsingSelector: @selector(compare:)];
    NSArray * objects = [revisionsByDay objectsForKeys: sortedKeys notFoundMarker: [NSNull null]];
    
    return objects;
}

- (void)updateRevisionsByDay:(NSMutableDictionary*)revisionsByDay withRevision:(WMFRevision*)revision {
    NSInteger distanceInDaysToDate = [revision daysFromToday];
    if (!revisionsByDay[@(distanceInDaysToDate)]) {
        revisionsByDay[@(distanceInDaysToDate)] = @[].mutableCopy;
    }
    
    NSMutableArray* revisionRowArray = revisionsByDay[@(distanceInDaysToDate)];
    [revisionRowArray addObject:revision];
}

@end

File: PageHistoryFetcher.swift
import Foundation
import AFNetworking
import Mantle


public class PageHistoryFetcher: NSObject {
    private let operationManager: AFHTTPSessionManager = {
        let manager = AFHTTPSessionManager.wmf_createDefaultManager()
        manager.responseSerializer = WMFApiJsonResponseSerializer()
        manager.requestSerializer = PageHistoryRequestSerializer()
        return manager
    }()

    public func fetchRevisionInfo(title: MWKTitle, requestParams: PageHistoryRequestParameters) -> AnyPromise {
        return AnyPromise(resolverBlock: { [weak self] (resolve) in
            guard let strongSelf = self else { return }
            strongSelf.operationManager.wmf_GETWithSite(title.site,
                                                        parameters: requestParams,
                                                        retry: nil,
                                                        success: { (operation, responseObject) in
                                                            guard let strongSelf = self, responseDict = responseObject as? [String: AnyObject] else { return }
                                                            MWNetworkActivityIndicatorManager.sharedManager().pop()
                                                            strongSelf.updatePagingState(responseDict)
                                                            resolve(strongSelf.parseSections(responseDict))
                                                            },
                                                        failure: { (operation, error) in
                                                                MWNetworkActivityIndicatorManager.sharedManager().pop()
                                                                resolve(error)
                                                        })
        })
    }
    
    //Mark: Paging
    private var continueKey: String?
    private var rvContinueKey: String?
    var batchComplete: Bool = false
    private var lastRevisionFromPreviousCall: WMFPageHistoryRevision?

    private func updatePagingState(responseDict: [String: AnyObject]) {
        if let continueInfo = responseDict["continue"] as? [String: AnyObject] {
            continueKey = continueInfo["continue"] as? String
            rvContinueKey = continueInfo["rvcontinue"] as? String
        }
        if responseDict["batchcomplete"] != nil {
            batchComplete = true
        }
    }
    
    //Mark: Data Parsing
    private typealias RevisionCurrentPrevious = (current: WMFPageHistoryRevision, previous: WMFPageHistoryRevision)
    private typealias RevisionsByDay = [Int: PageHistorySection]

    private func parseSections(responseDict: [String: AnyObject]) -> [PageHistorySection] {
        guard let pages = responseDict["query"]?["pages"] as? [String: AnyObject] else {
            assertionFailure("couldn't parse page history response")
            return []
        }
        
        var revisionsByDay = RevisionsByDay()
        for (_, value) in pages {
            let transformer = MTLJSONAdapter.arrayTransformerWithModelClass(WMFPageHistoryRevision.self)
            
            guard var revisions = transformer.transformedValue(value["revisions"]) as? [WMFPageHistoryRevision] else {
                assertionFailure("couldn't parse page history revisions")
                return []
            }
            if let leftoverRevisionFromPreviousCall = lastRevisionFromPreviousCall {
                revisions.insert(leftoverRevisionFromPreviousCall, atIndex: 0)
            }
            
            revisionsByDay = parse(revisions: revisions, existingRevisions: revisionsByDay)
            
            if let earliestRevision = revisions.last where earliestRevision.parentID == 0 {
                earliestRevision.revisionSize = earliestRevision.articleSizeAtRevision
                update(revisionsByDay: &revisionsByDay, revision: earliestRevision)
            } else {
                lastRevisionFromPreviousCall = revisions.last
            }
        }
        
        return revisionsByDay.keys.sort(<).flatMap() { revisionsByDay[$0] }
    }
    
    private func parse(revisions revisions: [WMFPageHistoryRevision], existingRevisions: RevisionsByDay) -> RevisionsByDay {
        return zip(revisions, revisions.dropFirst()).reduce(existingRevisions, combine: { (revisionsByDay, itemPair: RevisionCurrentPrevious) -> RevisionsByDay in
            var revisionsByDay = revisionsByDay
            
            itemPair.current.revisionSize = itemPair.current.articleSizeAtRevision - itemPair.previous.articleSizeAtRevision
            update(revisionsByDay:&revisionsByDay, revision: itemPair.current)
            
            return revisionsByDay
        })
    }
    
    private func update(inout revisionsByDay revisionsByDay: RevisionsByDay, revision: WMFPageHistoryRevision) {
        let distanceToToday = revision.daysFromToday()
        
        if let existingRevisionsOnCurrentDay = revisionsByDay[distanceToToday] {
            let sectionTitle = existingRevisionsOnCurrentDay.sectionTitle
            let items = existingRevisionsOnCurrentDay.items + [revision]
            revisionsByDay[distanceToToday] = PageHistorySection(sectionTitle: sectionTitle, items: items)
        } else {
            if let revisionDate = revision.revisionDate {
                let sectionTitle = NSDateFormatter.wmf_longDateFormatter().stringFromDate(revisionDate)
                let newSection = PageHistorySection(sectionTitle: sectionTitle, items: [revision])
                revisionsByDay[distanceToToday] = newSection
            }
        }
    }
}

public class PageHistoryRequestParameters: NSObject {
    let title: String
    var continueKey: String?
    var rvContinueKey: String?
    
    init(title: String) {
        self.title = title
    }
}

public class PageHistoryRequestSerializer: AFHTTPRequestSerializer {
    public override func requestBySerializingRequest(request: NSURLRequest, withParameters parameters: AnyObject?, error: NSErrorPointer) -> NSURLRequest? {
        guard let pageHistoryParameters = parameters as? PageHistoryRequestParameters else {
            assertionFailure("pagehistoryfetcher has incorrect parameter type")
            return nil
        }
        return super.requestBySerializingRequest(request, withParameters: serializedParams(pageHistoryParameters), error: error)
    }
    
    func serializedParams(requestParameters: PageHistoryRequestParameters) -> [String: AnyObject] {
        var params: [String: AnyObject] = [
            "action": "query",
            "prop": "revisions",
            "rvprop": "ids|timestamp|user|size|parsedcomment",
            "rvlimit": 51,
            "rvdir": "older",
            "titles": requestParameters.title,
            "continue": requestParameters.continueKey ?? "",
            "format": "json"
            //,"rvdiffto": -1 //Add this to fake out "error" api response.
        ]
        
        if let rvContinueKey = requestParameters.rvContinueKey {
            params["rvcontinue"] = rvContinueKey
        }
        
        return params
    }
}